common/src/main/java/io/netty/util/RedBlackTree.java

```diff
@@ -0,0 +1,525 @@
+/*
+ * Copyright 2013 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.util;
+
+import java.util.NoSuchElementException;
+
+public class RedBlackTree<Key extends Comparable<Key>> {
```

Comment:
Given that for the allocator the only version we need is using Long, although less reusable, why not using a long (unboxed) version instead?

Reply:
Do you mean using long instead of generics? I think it's better to make it reusable. For future use.

Reply:
I think @franz1981 is pushing for a second implementation LongRedBlackTree that uses primitive long.. @franz1981 is unboxing boxing really that expensive?

Reply:
Exactly, until value objects won't land on JDK, having pointer chasing/GC barriers and additional memory footprint (8 bytes Vs >= 28 bytes) means that unless we really know it will be used for other purposes it's often a good idea to use the primitive version, especially at this stage where such collection is just introduced. 
Anyway, how much we expect this to be used? In the hot path? How many different keys it contains for our use case? How's the size? It will be ofte modified or just read?
Knowing such things will let decide if we can save from be refactored with primitive keys if we want to be more pragmatic